python/lib/scheduler/dmod/scheduler/job/job.py

```diff
@@ -1230,221 +979,108 @@ def worker_data_requirements(self) -> List[List[DataRequirement]]:
-            "allocation_priority" : 0,
-            "job_id" : "12345678-1234-5678-1234-567812345678",
-            "rsa_key_pair" : {<serialized_representation_of_RsaKeyPair_obj>},
-            "status" : INIT:DEFAULT,
-            "last_updated" : "2020-07-10 12:05:45",
-            "allocations" : [...],
-            'data_requirements" : [...],
-            "partitioning" : { "partitions": [ ... ] },
-            "originating_request" : {<serialized_representation_of_originating_message>}
-        }
-
-        Returns
-        -------
-        dict
-            the representation of this instance as a dictionary or dictionary-like object (e.g., a JSON object)
-        """
-        dictionary = super().to_dict()
-        # To avoid this being messy, rely on the superclass's implementation and the returned dict, but remove the
-        # 'model_request' key/value, since this is contained within the originating serialized scheduler request
-        dictionary.pop('model_request')
-        dictionary['originating_request'] = self.originating_request.to_dict()
-        return dictionary
+    originating_request: SchedulerRequestMessage
+    """The original request that resulted in the creation of this job."""
+
+    class Config: # type: ignore
+        fields = {
+            # exclude `model_request` during serialization
+            "model_request": {"exclude": True}
+            }
+
+    # TODO: #pydantic_refactor - make sure we really no longer need the old factory_init_from_request class method
+
+    def __init__(self, job_request: SchedulerRequestMessage = None, **data):
+        if data:
+            # NOTE: in previous version of code, `model_request` was always a derived field.
+            # this allows `model_request` be separately specified
+            if "model_request" in data:
+                super().__init__(**data)
+                return
+
+            originating_request = data.get("originating_request")
+            if originating_request is None:
+                # this should fail, let pydantic handle that.
+                super().__init__(**data)
+                return
+
+            if isinstance(originating_request, SchedulerRequestMessage):
+                # inject
+                data["model_request"] = originating_request.model_request
+
+            data["model_request"] = originating_request.get("model_request")
+            super().__init__(**data)
+            return
+
+        # NOTE: consider refactoring this into `from_job_request` class method.
+        super().__init__(
+            cpu_count=job_request.cpus,
+            memory_size=job_request.memory,
+            model_request=job_request.model_request,
+            allocation_paradigm=job_request.allocation_paradigm,
+            originating_request=job_request,
+            )
+        # NOTE: this implicitly resets `last_updated` field
+        self.set_data_requirements(job_request.model_request.data_requirements)
```

Comment:
> it will set `data["model_request"]` twice, once assuming that `originating_request` is a `SchedulerRequestMessage` and the other assuming `originating_request` is a `Mapping`.

Yeah, I think that is a bug that I need to fix. I think your code is more readable. I will refactor what I have to resemble something closer to your suggestion. Thanks for that, @christophertubbs!

@aaraney #TODO

Notes:

```python
def __init__(self, job_request: SchedulerRequestMessage = None, **data):
    # NOTE: in previous version of code, `model_request` was always a derived field.
    # this allows `model_request` be separately specified
    if "model_request" in data:
        super().__init__(**data)
        return

    if data:
        originating_request = data.get("originating_request", dict())
        if isinstance(originating_request, SchedulerRequestMessage):
            data["model_request"] = originating_request.model_request
        else:
            data["model_request"] = originating_request.get("model_request")
        super().__init__(**data)
        return

    super().__init__(
            cpu_count=job_request.cpus,
            memory_size=job_request.memory,
            model_request=job_request.model_request,
            allocation_paradigm=job_request.allocation_paradigm,
            originating_request=job_request,
        )
    # NOTE: this implicitly resets `last_updated` field
    self.set_data_requirements(job_request.model_request.data_requirements)
```